todo/operations/csv_importer.py

```diff
@@ -0,0 +1,185 @@
+import codecs
+import csv
+import datetime
+import logging
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import Group
+
+from todo.models import Task, TaskList
+
+log = logging.getLogger(__name__)
+
+
+class CSVImporter:
+    """Core upsert functionality for CSV import, for re-use by `import_csv` management command, web UI and tests.
+    Supplies a detailed log of what was and was not imported at the end. See README for usage notes.
+    """
+
+    def __init__(self):
+        self.results = {
+            "errors": [],
+            "upserts": [],
+            "summaries": [],
+        }
+        self.line_count = 0
+        self.upsert_count = 0
+
+    def upsert(self, fileobj, as_string_obj=False):
+        """Expects a file *object*, not a file path. This is important because this has to work for both
+        the management command and the web uploader; the web uploader will pass in in-memory file
+        with no path!
+
+        Header row is:
+        Title, Group, Task List, Created Date, Due Date, Completed, Created By, Assigned To, Note, Priority
+        """
+
+        if as_string_obj:
+            # fileobj comes from mgmt command
+            csv_reader = csv.DictReader(fileobj)
+        else:
+            # fileobj comes from browser upload (in-memory)
+            csv_reader = csv.DictReader(codecs.iterdecode(fileobj, "utf-8"))
+
+        # DI check: Do we have expected header row?
+        header = csv_reader.fieldnames
+        expected = ['Title', 'Group', 'Task List', 'Created By', 'Created Date', 'Due Date', 'Completed', 'Assigned To', 'Note', 'Priority']
```

Comment:
what would you think about moving all validator snippets to separate functions, registering at the same time the expected field name (with a decorator) ? I think it would make `validate_row` less messy, and avoid explicitly listing fields

Reply:
Yeah, I was thinking the same. That one sort of grew organically I got lazy. Will fix.